### PR TITLE
Optimistic: dm-master support conflict DDL in optimistic mode.

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: ref #3787  (dm-master)

### What is changed and how it works?
- when master receive a conflict DDL, return SkipWaitRedirection operation
- when master receive same conflict DDL for all tables, return Resolved operation
- when master receive different conflict DDLs for different tables, return ConflictDetect operation

### Check List 

Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
Support conflict DDL in optimistic sharding mode.
```

### CODE REVIEW

**Comment by GMHDBJD** [comment]
/hold 
waiting for #3903

**Comment by lonng** [comment]
@GMHDBJD This pull request is too large we need to split it into some smaller pull requests.

**Comment by GMHDBJD** [comment]
> @GMHDBJD This pull request is too large we need to split it into some smaller pull requests.

This pr is base on #3903, it will be small after #3903 merged. 😂

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [inline dm/pkg/shardddl/optimism/lock.go:303]: These logic move to func TrySyncForOneDDL

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [inline dm/pkg/shardddl/optimism/lock.go:371]: these logic move to func checkAddDropColumns

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [inline dm/pkg/shardddl/optimism/lock.go:772]: This is the main logic to support conflict DDL

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [inline dm/pkg/shardddl/optimism/lock.go:1117]: main logic for redirect conflict DDL

File: dm/pkg/shardddl/optimism/lock_test.go
Comment (GMHDBJD) [inline dm/pkg/shardddl/optimism/lock_test.go:2367]: example of judge a conflict DDL

File: dm/pkg/shardddl/optimism/lock_test.go
Comment (GMHDBJD) [inline dm/pkg/shardddl/optimism/lock_test.go:2653]: results of different  DDLs

File: dm/pkg/shardddl/optimism/operation.go
Comment (GMHDBJD) [inline dm/pkg/shardddl/optimism/operation.go:44]: New operation for conflict DDL

File: dm/pkg/shardddl/optimism/lock.go
Comment (Ehco1996) [reply dm/pkg/shardddl/optimism/lock.go:371]: this func name is `checkAddDropColumn` 😬

File: dm/pkg/shardddl/optimism/lock.go
Comment (Ehco1996) [inline dm/pkg/shardddl/optimism/lock.go:719]: droped -> dropped

File: dm/pkg/shardddl/optimism/lock.go
Comment (Ehco1996) [inline dm/pkg/shardddl/optimism/lock.go:731]: just a question why we can mute joined and compared err?

File: dm/pkg/shardddl/optimism/lock.go
Comment (Ehco1996) [inline dm/pkg/shardddl/optimism/lock.go:747]: dose cmp <=0 means it a drop column operateion ? seesm in this branch, you not return the dropped column when err not hanppened

File: dm/pkg/shardddl/optimism/lock.go
Comment (Ehco1996) [inline dm/pkg/shardddl/optimism/lock.go:759]: why we both check `if cmp <=0` and `if cmp >=0`🤔

File: dm/pkg/shardddl/optimism/lock.go
Comment (Ehco1996) [inline dm/pkg/shardddl/optimism/lock.go:772]: perfer rename to `trySyncForOneDDL `, because no outer caller and we not use lock in this func

File: dm/pkg/shardddl/optimism/lock.go
Comment (Ehco1996) [inline dm/pkg/shardddl/optimism/lock.go:1121]: i'm a bit concerned about the performance issue of etcd here, it's a triple neste loop 😭, can we batch these ops?

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:107]: We now calculate pre join every time when receive a info(L787), because pre join may has error.

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:222]: We do not update the DropColumn if any errors occur.

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:225]: It is updated in TrySyncForOneDDL

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:731]: This error won't happen because we only call it when no conflict L262

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:747]: cmp <= 0 means a ADD column operation(oldJoin <= newJoin), so there is no dropped column. This used to check
```
source1: add column a varchar(4)
source2: add column a varchar(5)
```
then varchar(5) will be ignored(column already exist)

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:759]: ``
source1: add column a varchar(4) -- info1
source2: add column a varchar(5) -- info2
```
when receive info2, cmp will be 0, we need check for error

```
source1: drop column a -- info1
source2: drop column a -- info2
```
when reveice info1, cmp will be 0, we need check for drop column

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:1121]: later pr with #3829

File: dm/pkg/shardddl/optimism/lock.go
Comment (lichunzhu) [inline dm/pkg/shardddl/optimism/lock.go:1123]: Should put DDLs for resolved operation

File: dm/pkg/shardddl/optimism/lock_test.go
Comment (Ehco1996) [inline dm/pkg/shardddl/optimism/lock_test.go:2322]: please use testify when write brand new test case 😭😭😭😭

**Review by Ehco1996 — COMMENTED** [review]
restLGTM

File: dm/pkg/shardddl/optimism/lock_test.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock_test.go:2322]: This pr was written before the policy started. I don't want to change it. xia ci yi ding.

**Review by Ehco1996 — APPROVED** [review]
LGTM

File: dm/pkg/shardddl/optimism/lock_test.go
Comment (lichunzhu) [inline dm/pkg/shardddl/optimism/lock_test.go:2603]: Do we need a check for redirect operations?

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:224]: address in [7ace980](https://github.com/pingcap/tiflow/pull/4297/commits/7ace980489a78ce75cc8e638a21d36925e31cd1d)

File: dm/pkg/shardddl/optimism/lock.go
Comment (lichunzhu) [inline dm/pkg/shardddl/optimism/lock.go:806]: It looks like this could be put after L813 🤔

**Review by lichunzhu — APPROVED** [review]
LGTM

File: dm/pkg/shardddl/optimism/lock.go
Comment (GMHDBJD) [reply dm/pkg/shardddl/optimism/lock.go:806]: No. L808-L809 update the prev_tables, we should calculate oldJoined before it.

**Pull request merged**
